be/src/vec/aggregate_functions/aggregate_function_collect.h

```diff
@@ -203,27 +279,36 @@ class AggregateFunctionCollect final
     }
 
     DataTypePtr get_return_type() const override {
-        return std::make_shared<DataTypeArray>(make_nullable(_argument_type));
+        return std::make_shared<DataTypeArray>(make_nullable(return_type));
     }
 
+    bool allocates_memory_in_arena() const override { return ENABLE_ARENA; }
+
     void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
              Arena* arena) const override {
-        assert(!columns[0]->is_null_at(row_num));
-        if constexpr (alloc_memory_in_arena) {
-            this->data(place).add(*columns[0], row_num, arena);
+        auto& data = this->data(place);
+        if constexpr (HAS_LIMIT) {
+            data.max_size =
+                    (UInt64) static_cast<const ColumnInt32*>(columns[1])->get_element(row_num);
+            if (data.size() >= data.max_size) {
+                return;
+            }
+        }
+        if constexpr (ENABLE_ARENA) {
+            data.add(*columns[0], row_num, arena);
         } else {
-            this->data(place).add(*columns[0], row_num);
+            data.add(*columns[0], row_num);
         }
     }
 
-    void reset(AggregateDataPtr place) const override { this->data(place).reset(); }
-
     void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs,
                Arena* arena) const override {
-        if constexpr (alloc_memory_in_arena) {
-            this->data(place).merge(this->data(rhs), arena);
+        auto& data = this->data(place);
+        auto& rhs_data = this->data(rhs);
+        if constexpr (ENABLE_ARENA) {
+            data.merge(rhs_data, HAS_LIMIT, arena);
```

Comment:
whether the `HAS_LIMIT` pass to  dataSet a template, so doing merge work like
`if constexpr has_limit {
 }.....`

Reply:
It does seem more graceful, I'll try to enhance in this way.